## Tiền xử lý dữ liệu

In [2]:
import polars as pl 
import matplotlib.pyplot as plt

df = pl.scan_csv("output/20250715-ketquathi-ct2018a.csv")

df = df.with_columns([
    pl.col("SOBAODANH").cast(pl.Utf8).str.zfill(8).alias("SOBAODANH"),
    
    *[
    pl.col(col).cast(pl.Float64).alias(col)
        for col in df.collect_schema().names()
        if col not in ["STT", "SOBAODANH", "Mã tỉnh thành"]
    ]
])

df.collect().sample(n=5)

shape: (5, 21)
┌─────────┬───────────┬──────┬──────┬───┬──────┬──────┬──────┬───────────────┐
│ STT     ┆ SOBAODANH ┆ Toán ┆ Văn  ┆ … ┆ Đức  ┆ Pháp ┆ Nga  ┆ Mã tỉnh thành │
│ ---     ┆ ---       ┆ ---  ┆ ---  ┆   ┆ ---  ┆ ---  ┆ ---  ┆ ---           │
│ i64     ┆ str       ┆ f64  ┆ f64  ┆   ┆ f64  ┆ f64  ┆ f64  ┆ i64           │
╞═════════╪═══════════╪══════╪══════╪═══╪══════╪══════╪══════╪═══════════════╡
│ 1043250 ┆ 55010475  ┆ 5.25 ┆ 6.75 ┆ … ┆ null ┆ null ┆ null ┆ 55            │
│ 870611  ┆ 45004629  ┆ 4.5  ┆ 7.75 ┆ … ┆ null ┆ null ┆ null ┆ 45            │
│ 790589  ┆ 40006399  ┆ 2.45 ┆ 6.5  ┆ … ┆ null ┆ null ┆ null ┆ 40            │
│ 229691  ┆ 03012038  ┆ 3.6  ┆ 5.25 ┆ … ┆ null ┆ null ┆ null ┆ 3             │
│ 231720  ┆ 03014069  ┆ 7.5  ┆ 8.0  ┆ … ┆ null ┆ null ┆ null ┆ 3             │
└─────────┴───────────┴──────┴──────┴───┴──────┴──────┴──────┴───────────────┘

In [3]:
df.collect_schema().names()

['STT',
 'SOBAODANH',
 'Toán',
 'Văn',
 'Lí',
 'Hóa',
 'Sinh',
 'Tin học',
 'Công nghệ công nghiệp',
 'Công nghệ nông nghiệp',
 'Sử',
 'Địa',
 'Giáo dục kinh tế và pháp luật',
 'Anh',
 'Nhật',
 'Trung',
 'Hàn',
 'Đức',
 'Pháp',
 'Nga',
 'Mã tỉnh thành']

In [4]:
df.describe()

shape: (9, 22)
┌────────────┬───────────────┬───────────┬────────────┬───┬────────────┬────────────┬────────────┬───────────────┐
│ statistic  ┆ STT           ┆ SOBAODANH ┆ Toán       ┆ … ┆ Đức        ┆ Pháp       ┆ Nga        ┆ Mã tỉnh thành │
│ ---        ┆ ---           ┆ ---       ┆ ---        ┆   ┆ ---        ┆ ---        ┆ ---        ┆ ---           │
│ str        ┆ f64           ┆ str       ┆ f64        ┆   ┆ f64        ┆ f64        ┆ f64        ┆ f64           │
╞════════════╪═══════════════╪═══════════╪════════════╪═══╪════════════╪════════════╪════════════╪═══════════════╡
│ count      ┆ 1.131136e6    ┆ 1131136   ┆ 1.131136e6 ┆ … ┆ 150.0      ┆ 408.0      ┆ 60.0       ┆ 1.131136e6    │
│ null_count ┆ 0.0           ┆ 0         ┆ 0.0        ┆ … ┆ 1.130986e6 ┆ 1.130728e6 ┆ 1.131076e6 ┆ 0.0           │
│ mean       ┆ 565568.5      ┆ null      ┆ 4.762109   ┆ … ┆ 7.816667   ┆ 8.376838   ┆ 8.941667   ┆ 26.881627     │
│ std        ┆ 326530.981383 ┆ null      ┆ 1.703135   ┆ … ┆ 1.641448   ┆ 1.46792    ┆ 1.156564   ┆ 19.239681     │
│ min        ┆ 1.0           ┆ 01000001  ┆ 0.0        ┆ … ┆ 3.5        ┆ 3.25       ┆ 5.25       ┆ 1.0           │
│ 25%        ┆ 282785.0      ┆ null      ┆ 3.5        ┆ … ┆ 6.75       ┆ 7.75       ┆ 8.75       ┆ 8.0           │
│ 50%        ┆ 565569.0      ┆ null      ┆ 4.6        ┆ … ┆ 8.25       ┆ 8.75       ┆ 9.25       ┆ 27.0          │
│ 75%        ┆ 848352.0      ┆ null      ┆ 5.85       ┆ … ┆ 9.0        ┆ 9.5        ┆ 9.75       ┆ 43.0          │
│ max        ┆ 1.131136e6    ┆ 64007242  ┆ 10.0       ┆ … ┆ 10.0       ┆ 10.0       ┆ 10.0       ┆ 64.0          │
└────────────┴───────────────┴───────────┴────────────┴───┴────────────┴────────────┴────────────┴───────────────┘

In [5]:
necessary_df_bool = df.drop(["STT", "SOBAODANH", "Toán", "Văn", "Mã tỉnh thành"])
    
necessary_df_bool = necessary_df_bool.select(
    [pl.col(c).is_not_null().alias(c) for c in necessary_df_bool.collect_schema().names()]
).collect().to_pandas()

necessary_df_bool.head()

Lí    Hóa   Sinh  Tin học  Công nghệ công nghiệp  Công nghệ nông nghiệp  ...   Nhật  Trung    Hàn    Đức   Pháp    Nga
0  False   True   True    False                  False                  False  ...  False  False  False  False  False  False
1   True   True  False    False                  False                  False  ...  False  False  False  False  False  False
2   True  False  False    False                  False                  False  ...  False  False  False  False  False  False
3   True   True  False    False                  False                  False  ...  False  False  False  False  False  False
4  False  False  False    False                  False                  False  ...  False  False  False  False  False  False

[5 rows x 16 columns]

## Luật kết hợp (Apriori)

In [14]:
from mlxtend.frequent_patterns import apriori, association_rules, fpgrowth
import matplotlib.pyplot as plt

In [15]:
combine = apriori(
    necessary_df_bool, 
    min_support = 0.025, 
    use_colnames = True
)

rules = association_rules(
    combine, 
    metric="confidence", 
    min_threshold=0.3
)

In [16]:
rules[['antecedents', 'consequents', 'support', 'confidence', 'lift', 'leverage']]

antecedents consequents   support  confidence      lift  leverage
0                            (Hóa)        (Lí)  0.143401    0.675478  2.198102  0.078162
1                             (Lí)       (Hóa)  0.143401    0.466647  2.198102  0.078162
2                             (Lí)       (Anh)  0.129924    0.422792  1.359209  0.034336
3                            (Anh)        (Lí)  0.129924    0.417686  1.359209  0.034336
4                           (Sinh)       (Hóa)  0.039751    0.643308  3.030248  0.026633
5                             (Sử)       (Địa)  0.262924    0.617925  1.466943  0.083691
6                            (Địa)        (Sử)  0.262924    0.624177  1.466943  0.083691
7  (Giáo dục kinh tế và pháp luật)       (Địa)  0.096741    0.444101  1.054288  0.004981

## Luật kết hợp (FPGrowth)

In [17]:
combine = fpgrowth(
    necessary_df_bool, 
    min_support = 0.025, 
    use_colnames = True
)

rules = association_rules(
    combine, 
    metric="confidence", 
    min_threshold=0.3
)

In [18]:
rules[['antecedents', 'consequents', 'support', 'confidence', 'lift', 'leverage']]

antecedents consequents   support  confidence      lift  leverage
0                            (Hóa)        (Lí)  0.143401    0.675478  2.198102  0.078162
1                             (Lí)       (Hóa)  0.143401    0.466647  2.198102  0.078162
2                           (Sinh)       (Hóa)  0.039751    0.643308  3.030248  0.026633
3                             (Lí)       (Anh)  0.129924    0.422792  1.359209  0.034336
4                            (Anh)        (Lí)  0.129924    0.417686  1.359209  0.034336
5                             (Sử)       (Địa)  0.262924    0.617925  1.466943  0.083691
6                            (Địa)        (Sử)  0.262924    0.624177  1.466943  0.083691
7  (Giáo dục kinh tế và pháp luật)       (Địa)  0.096741    0.444101  1.054288  0.004981

## Kết luận

1. **Nhóm Tự nhiên**

- **Hóa ↔ Lí**
    - Support ~14% → khá nhiều thí sinh chọn cả hai.
    - Confidence Hóa → Lí = 68% (rất cao).
    - Lift = 2.2 → mối liên hệ mạnh gấp đôi so với ngẫu nhiên.
    - 👉 Đây là cặp môn đi cùng nhau rõ rệt, phản ánh truyền thống khối A.

- **Sinh → Hóa**
    - Confidence = 64%, Lift = 3.0 → cực kỳ mạnh.
    - 👉 Thí sinh chọn Sinh gần như chắc chắn chọn Hóa, phản ánh khối B (Toán–Hóa–Sinh).
    - Ngược lại, Hóa → Sinh không mạnh bằng, vì Hóa còn gắn với khối A.

2. **Nhóm Xã hội**

- **Sử ↔ Địa**
    - Support ~26%, Confidence ~62%, Lift ~1.47 → mối liên hệ khá rõ.
    - 👉 Đây là cặp môn đặc trưng của khối C (Văn–Sử–Địa).

3. **Các mối liên hệ khác**

- **Anh ↔ Lí**
    - Confidence ~42%, Lift ~1.36 → có xu hướng, nhưng không mạnh bằng các cặp truyền thống.
    - 👉 Có thể phản ánh một số thí sinh chọn kết hợp khối D với môn tự nhiên.

- **GD Kinh tế & Pháp luật → Địa**
    - Lift ≈ 1.05
    - 👉 gần như ngẫu nhiên, không đáng chú ý.


==> **KẾT LUẬN**

- Đáng chú ý nhất: Nhóm Tự nhiên (Hóa–Lí–Sinh) và Nhóm Xã hội (Sử–Địa).
- Một chiều đặc trưng: Sinh → Hóa.
- Hai chiều mạnh: Hóa ↔ Lí, Sử ↔ Địa.
- Không đáng chú ý: GD Kinh tế & Pháp luật → Địa.

In [19]:
!pip freeze > requirements.txt